kernel/src/hil/adc.rs

```diff
@@ -1,38 +1,80 @@
+/// Interfaces for analog to digital converter peripherals.
 use returncode::ReturnCode;
 
-/// Trait for handling callbacks from ADC module.
-pub trait Client {
-    /// Called when a sample is ready.
-    fn sample_done(&self, sample: u16);
-}
 
-/// Simple interface for reading a single ADC sample on any channel.
-pub trait AdcSingle {
+// *** Interfaces for low-speed, single-sample ADCs ***
+
+/// Simple interface for reading an ADC sample on any channel.
+pub trait Adc {
+    /// The chip-dependent type of an ADC channel.
+    type Channel;
+
     /// Initialize must be called before taking a sample.
-    /// Returns true on success.
     fn initialize(&self) -> ReturnCode;
 
     /// Request a single ADC sample on a particular channel.
-    /// Returns true on success.
-    fn sample(&self, channel: u8) -> ReturnCode;
-    fn cancel_sample(&self) -> ReturnCode;
+    /// Used for individual samples that have no timing requirements.
+    fn sample(&self, channel: &Self::Channel) -> ReturnCode;
+
+    /// Request repeated ADC samples on a particular channel.
+    /// Callbacks will occur at the given frequency with low jitter and can be set to any frequency
+    /// supported by the chip implementation. However callbacks may be limited based on how quickly
+    /// the system can service individual samples, leading to missed samples at high frequencies.
+    fn sample_continuous(&self, channel: &Self::Channel, frequency: u32) -> ReturnCode;
+
+    /// Stop a sampling operation.
+    /// Can be used to stop simple or high-speed samples. No further callback will occur in simple
+    /// mode. The `sampling_complete` callback will occur in high-speed mode.
+    fn stop_sampling(&self) -> ReturnCode;
```

Comment:
These issues of reclaiming buffers are why I originally proposed the callback approach; the idea is that cancelling is itself asynchronous. You're not fully cancelled until you get the last sampling callback, which is an event. That way you have the same buffer management approach for handling completion of a sample and cancellation.

How about implicit cancellation by not providing buffers? I.e., you could say that cancel() only succeeds when the driver has no buffers left. On one hand, this makes the client logic a little more complicated. On the other, it sets a clear contract, that once you provide a buffer it will be filled. I'd be wary of this for slow continuous sampling (long long latency) but for high speed sampling it seems like it could be fine?